### DISCUSSION TITLE
Shipping an official DOM testing utility along with Next.js

### DISCUSSION THREAD

**Discussion by toomuchdesign** [vercel/next.js#19075] [Ideas]
Implementing integration tests for Next.js apps currently involves relying on browser e2e tests. That makes sense since we want to simultaneously test routes, SSR server and page component all together.

I recently made an attempt to **carry out Next.js integration tests using only DOM tests** with quite a promising outcome. The testing scope is of course more limited then browser tests and supports DOM testing page components along with file-system routing, fetching methods and custom App component.  

In order to do so I simplistically re-implement a few Next.js internal behaviour to be able to get the React element matching a specific route/page and test it with React Testing Library like:

```js
const { render } = await getPage({
  route: '/blog/1',
});

render();
expect(screen.getByText('Blog')).toBeInTheDocument();

fireEvent.click(screen.getByText('Link'));
await screen.findByText('Linked page');
```
The utility takes care of:

- **resolving** provided **routes** into the matching page component
- calling Next.js data **fetching methods** (`getServerSideProps`, `getInitialProps` or `getStaticProps`)
- set up the expected **mocked next/router provider**
- wrapping page with **custom _app component**
- **instantiating page component** with the expected props
- Emulate client side navigation via `Link`, `router.push`, `router.replace`

I published this **proof of concept** [here](https://github.com/toomuchdesign/next-page-tester), and I'm wondering whether a similar utility could be in future considered to be shipped as an **"official" Next.js DOM testing tool**.

Thanks for the great job you're all doing here.

**Comment by balazsorban44** [comment]
Related issue: https://github.com/vercel/next.js/issues/10967 (@toomuchdesign saw your name in the comments there as well)

If not the Next.js team, but maybe @testing-library would accept it into their list of testing libraries and it could also get some attention there for further improvement/new ideas. I think you should ask them as well.

**Reply by toomuchdesign** [reply]
Thanks for your suggestion, @balazsorban44.
After I wrote next-page-tester I realised how a tool like this would benefit from being officially supported by Next.js in terms of adoption and dependability.

I personally think Next.js as a framework would also benefit from such a cheap and reliable testing technique.

I'll consider knocking Testing library's door, even though this utility is testing tool agnostic.
